app/src/main/java/com/sopt/now/presentation/Login/LoginViewModel.kt

```diff
@@ -0,0 +1,58 @@
+package com.sopt.now.presentation.Login
+
+
+import androidx.lifecycle.MutableLiveData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.sopt.now.domain.AuthRepository
+import com.sopt.now.domain.model.AuthData
+import kotlinx.coroutines.launch
+import retrofit2.HttpException
+
+class LoginViewModel(
+    private val authRepository: AuthRepository
+) : ViewModel() {
+    val loginResult: MutableLiveData<Boolean> = MutableLiveData()
```

Comment:
```suggestion
    private val _loginResult: MutableLiveData<Boolean> = MutableLiveData()
    val loginResult: LiveData<Boolean> get() = _loginResult
```

그런데 여기서 수정해야 할 부분이 하나 더 존재합니다. 바로 MutableLiveData를 구독하는 부분인데요, Kotlin에서는 기본적으로 불변성을 지향합니다. ViewModel 내에서는 서버통신 등의 발행객체의 내부 상태를 변경해야하는 일이 있지만, **외부 객체(Activity, Fragment)에서는 이를 변경할 이유도 없고 오히려 변경한다면 예상치 못한 일들(SideEffect)이 발생할 수도 있습니다.**

따라서 이런 발행객체를 사용하는 경우 **상태를 변경할 수 있는 객체는 private으로, 상태를 공개하는 객체는 불변객체로 공개해야합니다.** 이런 기법을 **`Backing Property`**라고 합니다.

추가로 get() = 를 사용하지 않고 = 으로 값을 바로 할당하는 방법이 있는데 두 코드의 차이점이 무엇인지 찾아보고 댓글로 남겨주세요!
```kotlin
    val loginResult: LiveData<Boolean> get() = _loginResult
    val loginResult: LiveData<Boolean> = _loginResult
```

Reply:
내부에서 값을 변경할때는 _loginResult.value = 값
이런식으로 사용하시면 됩니다!

외부에서는 loginResult를 사용하구요!

Reply:
val loginResult: LiveData<Boolean> get() = _loginResult 를 사용하면 _loginResult가 변경될때마다 loginResult에 변경된 값을 넣어줍니다. 동적으로 값을 가져올 수 있다는 장점이 있습니다.
 
  val loginResult: LiveData<Boolean> = _loginResult
 _loginResult의 값을 초기화 시점에 loginResult에 할당합니다.

Reply:
[MutableLiveData]
LiveData는 Observer 패턴을 따르며 데이터의 변경이 일어날 때마다 콜백을 받습니다.
이때, LiveData의 값을 변경하게 해주는 함수는 setValue와 postValue가 있습니다.
두 함수는 protected로 되어있기에, 외부에서 값을 변경해주기 위해서는 LiveData를 상속받는 MutableLiveData를 사용해야 합니다. 

                    this@LoginViewModel.userId.value =it
                    loginResult.value = true

backing property를 사용하여 loginResult를 바꾸니 위의 부분에 문제가 생겼습니다.
MutableLiveData의 value는 메인 스레드에서만 사용할 수 있으므로 백그라운드에서 안전하게 값을 보낼 수 있는 postValue를 사용합니다.